src/gui/glRenderer.js

```diff
@@ -1,476 +1,753 @@
+                        case WorkerGLProxy.STATE_FRAME:
+                            this.serverState = WorkerGLProxy.STATE_FRAME_DONE;
+                            break;
+                        case WorkerGLProxy.STATE_CONTEXT_RESTORED:
+                            this.serverState = WorkerGLProxy.STATE_CONTEXT_RESTORED_DONE;
+                            break;
+                        default:
+                            console.error("received end frame in wrong state");
+                            break;
+                    }
+                    // make sure we stop waiting form more command buffers when the client indicates the work for this frame is done
+                    if (this.frameEndListener) {
+                        this.frameEndListener(true);
+                    }
+                    return;
                 }
-                let isReturnToDefault =
-                    (this.lastCapabilities[capaId] && message.name === 'enable') ||
-                    ((!this.lastCapabilities[capaId]) && message.name === 'disable');
-                if (isReturnToDefault) {
-                    delete this.lastCapabilities[capaId];
+                break;
+            default:
+                break;
+        }
+        switch (this.serverState) {
+            case WorkerGLProxy.STATE_FRAME:
+            case WorkerGLProxy.STATE_BOOTSTRAP:
+            case WorkerGLProxy.STATE_CONTEXT_RESTORED:
+                // if the message is none of the above, it's a command buffer. notify the commandbuffermanager about the received command buffer.
+                this.buffer.onCommandBufferReceived(message);
+                if (Atomics.load(this.synclock, 0) === 0) {
+                    switch (this.serverState) {
+                        case WorkerGLProxy.STATE_BOOTSTRAP:
+                            this.serverState = WorkerGLProxy.STATE_BOOTSTRAP_SYNC;
+                            break;
+                        case WorkerGLProxy.STATE_FRAME:
+                            this.serverState = WorkerGLProxy.STATE_FRAME_SYNC;
+                            break;
+                        case WorkerGLProxy.STATE_CONTEXT_RESTORED:
+                            this.serverState = WorkerGLProxy.STATE_CONTEXT_RESTORED_SYNC;
+                            break;
+                        default:
+                            break;
+                    }
+                }
+                break;
+            default:
+                break;
+        }
+    }
+
+    /**
+     * Executes an WebGL command on the server context and does basi error checking when debugging is enabled
+     * @param {string} name The function to call
+     * @param {Array<any>} args contains the internal arguments used by the WebGL instance (Handles translated to internal objects)
+     * @param {Array<any>} displayArgs contains the arguments received from the client (Handles etc...)
+     * @returns {any}
+     */
+    executeGL(name, args, displayArgs) {
```

Comment:
What's the purpose of displayArgs here? It doesn't appear to be used other than debugging

Reply:
it's only used for debugging. Because handles are resolved, I used it to display the original handles that were used, since that's what's being used on the client side when the command was given/added to the commandbuffer.
Let me clearify the description to make this more clear.